### PR TITLE
feat(explore): Allow using time formatter on temporal columns in data table

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
Enable switching between displaying epoch (original value) and time formatted value (format: YYYY-mm-dd HH:MM:SS).

Temporal columns in data table (tabs "Data" and "Samples") now have a clickable "gear" icon, which opens a popover that allows switching between 2 display modes. 

Selected display mode is specific to each column - user can select different display modes for each temporal column. The selection is also specific to datasource, meaning if there are temporal columns that have the same name in 2 datasources, those columns can have different display modes.

Selection is saved in local storage - when user reloads the page, changes hart config etc., it will be persisted.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/15073128/152353312-3abe36c5-1f23-4e72-8afc-991471b7722f.mov

### TESTING INSTRUCTIONS
1. Create a chart with some temporal columns
2. Open data table below the chart - temporal columns should have "gear" icons on their left sides
3. Click the icon and select "Formatted time"
4. Verify that the displayed values for that column have changed from long numbers to human readable dates
5. Reload the page, verify that your selections have been saved
6. Switch between datasources, verify that the selections are specific for each datasource
7. Perform the above testing instructions for charts using V1 API and legacy API

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

https://app.shortcut.com/preset/story/33355/explore-south-table-timestamp-column-in-data-panel-did-not-show-correctly

CC @rusackas @jinghua-qa @kasiazjc

### CODE REVIEW

File: superset/views/core.py [hunk 1]
```diff
@@ -449,10 +449,12 @@ def get_raw_results(self, viz_obj: BaseViz) -> FlaskResponse:
         payload = viz_obj.get_df_payload()
         if viz_obj.has_error(payload):
             return json_error_response(payload=payload, status=400)
-        return self.json_response({"data": payload["df"].to_dict("records")})
+        return self.json_response(
+            {"data": payload["df"].to_dict("records"), "colnames": payload["colnames"]}
```
Comment (villebro) [inline superset/views/core.py:453]: let's also add "coltypes" here so we get this feature working for legacy charts 👍

File: superset/viz.py [hunk 2]
```diff
@@ -258,8 +258,11 @@ def get_samples(self) -> List[Dict[str, Any]]:
                 "to_dttm": None,
             }
         )
-        df = self.get_df_payload(query_obj)["df"]  # leverage caching logic
-        return df.to_dict(orient="records")
+        payload = self.get_df_payload(query_obj)  # leverage caching logic
+        return {
+            "data": payload["df"].to_dict(orient="records"),
+            "colnames": payload["colnames"],
```
Comment (villebro) [inline superset/viz.py:264]: here too (requires small change in `viz.py`)

**Review by villebro — COMMENTED** [review]
First pass comments. Works really nicely! A minor wording change proposal + I recommend also adding it to legacy charts due to the minor needed changes.

File: superset-frontend/src/explore/components/DataTableControl/index.tsx [hunk 3]
```diff
@@ -97,6 +114,126 @@ export const RowCount = ({
...
+  onChange: any;
+  value: FormatPickerValue;
+}) => (
+  <Radio.Group value={value} onChange={onChange}>
+    <Space direction="vertical">
+      <Radio value={FormatPickerValue.epoch}>{t('Epoch')}</Radio>
+      <Radio value={FormatPickerValue.formatted}>{t('Formatted date')}</Radio>
+    </Space>
+  </Radio.Group>
+);
```
[the remark this answers is not in the document]
Comment (kgabryje) [reply superset-frontend/src/explore/components/DataTableControl/index.tsx]: Agreed on the first one - I changed "Epoch" to "Original value". However, as for "Formatted date", I think in this context user shouldn't confuse the word "date" with db type `DATE` - it feels more natural and intuitive to call it "Formatted date". I'm open to more arguments and suggestions though 🙂

File: superset/views/core.py (hunk 1 above)
Comment (kgabryje) [reply superset/views/core.py]: Done!

File: superset-frontend/src/explore/components/DataTableControl/index.tsx [hunk 4]
```diff
@@ -97,6 +114,128 @@ export const RowCount = ({
   />
 );
 
+enum FormatPickerValue {
+  formatted,
+  epoch,
```
[the remark this answers is not in the document]
Comment (kgabryje) [reply superset-frontend/src/explore/components/DataTableControl/index.tsx]: Done!

File: superset-frontend/src/explore/components/DataTableControl/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (villebro) [reply superset-frontend/src/explore/components/DataTableControl/index.tsx]: Sounds good - the "Formatted date" comment was mostly me trying to sound clever 😆

**Review by villebro — COMMENTED** [review]
Two last comments, other than that LGTM!

**Comment by zhaoyongjie** [comment]
Thanks for your hard work. Before I reviewed the code, I did some experiments. 

1. Let us use this SQL(PG) to make a virtual dataset, then open it in explore.
```
select
  '2022-01-03 00:00:00'::TIMESTAMP as datetime0,
  '2022-02-07 00:00:00'::TIMESTAMP as datetime1
UNION ALL
  SELECT '2022-01-04 00:00:00'::TIMESTAMP,
  '2022-02-08 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-05 00:00:00'::TIMESTAMP,
  '2022-02-09 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-06 00:00:00'::TIMESTAMP,
  '2022-02-10 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-07 00:00:00'::TIMESTAMP,
  '2022-02-11 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-08 00:00:00'::TIMESTAMP,
  '2022-02-12 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-09 00:00:00'::TIMESTAMP,
  '2022-02-13 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-10 00:00:00'::TIMESTAMP,
  '2022-02-14 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-11 00:00:00'::TIMESTAMP,
  '2022-02-15 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-12 00:00:00'::TIMESTAMP,
  '2022-02-16 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-13 00:00:00'::TIMESTAMP,
  '2022-02-17 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-14 00:00:00'::TIMESTAMP,
  '2022-02-18 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-15 00:00:00'::TIMESTAMP,
  '2022-02-19 00:00:00'::TIMESTAMP
UNION ALL
  SELECT '2022-01-16 00:00:00'::TIMESTAMP,
  '2022-02-20 00:00:00'::TIMESTAMP
```

2. Select *table*, put datatime0 and datetime1 to **groupby control**, put **count(*)** to metric, select **month** grain

3. viz result and query result is different.
image

This problem was not actually introduced by this PR, but it is a long-standing one.

let us check SQL, we can see time function only apply on **main time column**, 

```
SELECT DATE_TRUNC('month', datetime0) AS datetime0,
       datetime1 AS datetime1,
       count(*) AS count
FROM
  (select '2022-01-03 00:00:00'::TIMESTAMP as datetime0,
          '2022-02-07 00:00:00'::TIMESTAMP as datetime1
   UNION ALL SELECT '2022-01-04 00:00:00'::TIMESTAMP,
                    '2022-02-08 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-05 00:00:00'::TIMESTAMP,
                    '2022-02-09 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-06 00:00:00'::TIMESTAMP,
                    '2022-02-10 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-07 00:00:00'::TIMESTAMP,
                    '2022-02-11 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-08 00:00:00'::TIMESTAMP,
                    '2022-02-12 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-09 00:00:00'::TIMESTAMP,
                    '2022-02-13 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-10 00:00:00'::TIMESTAMP,
                    '2022-02-14 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-11 00:00:00'::TIMESTAMP,
                    '2022-02-15 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-12 00:00:00'::TIMESTAMP,
                    '2022-02-16 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-13 00:00:00'::TIMESTAMP,
                    '2022-02-17 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-14 00:00:00'::TIMESTAMP,
                    '2022-02-18 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-15 00:00:00'::TIMESTAMP,
                    '2022-02-19 00:00:00'::TIMESTAMP
   UNION ALL SELECT '2022-01-16 00:00:00'::TIMESTAMP,
                    '2022-02-20 00:00:00'::TIMESTAMP) AS virtual_table
GROUP BY DATE_TRUNC('month', datetime0),
         datetime1
ORDER BY count DESC
LIMIT 1000;
```

**Comment by villebro** [comment]
@zhaoyongjie the reason for this is that existing users are currently relying on the timegrain only applying to the primary temporal column. See the discussion here: #17239 We're actually currently exploring the possibility of introducing a time grain option in the column control, after which we will no longer be restricted to only having a single time grained temporal column.

**Comment by zhaoyongjie** [comment]
> @zhaoyongjie the reason for this is that existing users are currently relying on the timegrain only applying to the primary temporal column. See the discussion here: #17239 We're actually currently exploring the possibility of introducing a time grain option in the column control, after which we will no longer be restricted to only having a single time grained temporal column.

Thanks for the mention! I applied time format in the customize panel to resolve the inconsistency.

File: superset-frontend/src/explore/reducers/exploreReducer.js [hunk 5]
```diff
@@ -236,7 +237,46 @@ export default function exploreReducer(state = {}, action) {
...
+      const { datasourceId, columnName } = action;
+
+      if (Array.isArray(newTimeFormattedColumns[action.datasourceId])) {
+        newTimeFormattedColumns[datasourceId] = [
+          ...newTimeFormattedColumns[datasourceId],
+          columnName,
+        ];
+      } else {
+        newTimeFormattedColumns[datasourceId] = [columnName];
+      }
```
Comment (zhaoyongjie) [inline superset-frontend/src/explore/reducers/exploreReducer.js:251]:
```suggestion
      if (Array.isArray(newTimeFormattedColumns[action.datasourceId]) && !newTimeFormattedColumns[action.datasourceId].includes(columnName)) {
        newTimeFormattedColumns[datasourceId].push(columnName)
       } else {
        newTimeFormattedColumns[datasourceId] = [columnName];
      }
```

**Review by zhaoyongjie — APPROVED** [review]

File: superset-frontend/src/explore/reducers/exploreReducer.js [hunk 6]
```diff
@@ -236,7 +237,46 @@ export default function exploreReducer(state = {}, action) {
...
+
+      if (Array.isArray(newTimeFormattedColumns[datasourceId])) {
+        newTimeFormattedColumns[datasourceId] = [
+          ...newTimeFormattedColumns[datasourceId].slice(0, columnIndex),
+          ...newTimeFormattedColumns[datasourceId].slice(columnIndex + 1),
+        ];
+
+        if (newTimeFormattedColumns[datasourceId].length === 0) {
+          delete newTimeFormattedColumns[datasourceId];
+        }
```
[the remark this answers is not in the document]
Comment (kgabryje) [reply superset-frontend/src/explore/reducers/exploreReducer.js:270]: Thanks for comments! According to [redux docs](https://redux.js.org/usage/structuring-reducers/immutable-update-patterns), we cannot mutate the state - only immutable operations like `slice` are allowed. Splice mutates the original array. Same goes for using `push` in your second comment.

[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset-frontend/src/explore/reducers/exploreReducer.js:270]: Thanks for the mention. we can use more readable code like this:
```
        [...newTimeFormattedColumns[datasourceId]].splice(columnIndex, 1);
```
or
```
        newTimeFormattedColumns[datasourceId].slice().splice(columnIndex, 1);
```

**Review by villebro — APPROVED** [review]
LGTM!

File: superset-frontend/src/explore/reducers/exploreReducer.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset-frontend/src/explore/reducers/exploreReducer.js]: Thanks for the improvement!

**Pull request merged**